### PR TITLE
os/board/rtl8721csm : Implementation of Wi-Fi scan with SSID

### PR DESCRIPTION
**Pull request by keerthanalex**
Modified code to parse the result of scan with SSID API and to pass the list to upper layer
Modified LWNL_REQ_WIFI_SCANAP case of netdev_wifi.c to pass user input to scan with SSID API

### CODE REVIEW

**Review by pillip8282 — CHANGES_REQUESTED** [review]
it print scan result with ssid successfully. however driver doesn't generate a callback t to upper layer.

File: os/board/rtl8721csm/src/component/os/tizenrt/rtk_netmgr.c [hunk 1]
```diff
@@ -382,9 +382,16 @@ trwifi_result_e wifi_netmgr_utils_deinit(struct netdev *dev)
 trwifi_result_e wifi_netmgr_utils_scan_ap(struct netdev *dev, trwifi_ap_config_s *config)
 {
 	trwifi_result_e wuret = TRWIFI_FAIL;
-	if(wifi_scan_networks(app_scan_result_handler, NULL ) != RTW_SUCCESS){
-		//ndbg("[RTK] [ERR] WiFi scan fail(%d)\n", ret);
-		return wuret;
+	if (config->ssid != NULL) {
```
[the remark this answers is not in the document]
Comment (keerthanalex) [reply os/board/rtl8721csm/src/component/os/tizenrt/rtk_netmgr.c:385]: > trwifi_scan_ap API can receive null config if it's requested full scan. So you'd better to check validation of config.

@pillip8282 when trwifi_scan_ap AP receives null config, we have else part to do the full scan

else {		if (wifi_scan_networks(app_scan_result_handler, NULL ) != RTW_SUCCESS) {			//ndbg("[RTK] [ERR] WiFi scan fail(%d)\n", ret);			return wuret;		}	}

[the remark this answers is not in the document]
Comment (pillip8282) [reply os/board/rtl8721csm/src/component/os/tizenrt/rtk_netmgr.c:385]: If config is null then config->ssid will access null->ssid. I think this is dangerous.

File: os/board/rtl8721csm/src/component/os/tizenrt/rtk_netmgr.c [hunk 2]
```diff
@@ -382,9 +382,16 @@ trwifi_result_e wifi_netmgr_utils_deinit(struct netdev *dev)
 trwifi_result_e wifi_netmgr_utils_scan_ap(struct netdev *dev, trwifi_ap_config_s *config)
 {
 	trwifi_result_e wuret = TRWIFI_FAIL;
-	if(wifi_scan_networks(app_scan_result_handler, NULL ) != RTW_SUCCESS){
-		//ndbg("[RTK] [ERR] WiFi scan fail(%d)\n", ret);
-		return wuret;
+	if (config->ssid != NULL) {
+		int scan_buf_len = 50;
+		if (wifi_scan_networks_with_ssid(NULL, NULL, scan_buf_len, config->ssid, config->ssid_length) != RTW_SUCCESS) {
+			return wuret;
```
Comment (jeongchanKim) [inline os/board/rtl8721csm/src/component/os/tizenrt/rtk_netmgr.c:388]: `wuret` is not set from the return of functions.
So I think this variable is not needed.
How about using like `return TRWIFI_FAIL` or `return `TRWIFI_SUCCESS`?

File: os/board/rtl8721csm/src/component/os/tizenrt/rtk_netmgr.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (keerthanalex) [reply os/board/rtl8721csm/src/component/os/tizenrt/rtk_netmgr.c:385]: > If config is null then config->ssid will access null->ssid. I think this is dangerous.

ok @pillip8282 , I'll include this check as well.

**Review by pillip8282 — CHANGES_REQUESTED** [review]
Please check that wi-fi driver doesn't pass the scan result to upper layer. comparing scan with ssid and scan (full).

**Comment by keerthanalex** [comment]
> Please check that wi-fi driver doesn't pass the scan result to upper layer. comparing scan with ssid and scan (full).

@pillip8282 , I'm working on this currently and will update with result handler.

**Comment by pillip8282** [comment]
> > Please check that wi-fi driver doesn't pass the scan result to upper layer. comparing scan with ssid and scan (full).
> 
> @pillip8282 , I'm working on this currently and will update with result handler.
@keerthanalex 
In that case please add label to let me know the current state of PR. Because I thought that this PR is final so I've tested it every time.

**Comment by keerthanalex** [comment]
@pillip8282 , I updated the PR according to suggestions and would like to highlight two issues.

**1) SSID with space: for eg: "JJ Residence"**
The driver works fine for SSID with space as below,

TASH>>[WT] wifi manager test!!
[WT]  T16 wait func signal
[WT] -->wm_scan
netdev_handle_wifi: T6 cmd(3) (0x10064ac4) (116)
ssid len:(12)
len = 26,       mac = 64 cc 22 75 3e 3a ,        rssi = -72,    sec = wpa/wpa2, wps password id = 255,  channel = 6,    **ssid = JJ Residence**
len = 0,
[WT] T17 wait signal

But the condition at "_wt_parse_scan" of "wm_test.c" file,  makes "opt->scan_specific = 0" and redirects it to full scan. Please fix this function to test SSID with space.

**2) For the case with scan_cnt=0:**
In case of Invalid SSID/SSID out of range, "wifi_scan_networks_with_ssid()" API returns "RTW_SUCCESS" to upper layer if the scan process is "success" though it cannot fetch the details(scan_cnt=0). 

For such scenario, we will get below result

TASH>>wm_test scan welcome
TASH>>[WT][T16] wifi manager test!!     wm_test_main:907
[WT][T16] wait func signal      wm_test_main:914
[WT][T17] -->   _wt_scan:561
wifi_manager_scan_specific_ap: 197 [WM][INFO]   -->
wifimgr_handle_request: 621 [WM][INFO]  handle request state(STA_DISCONNECTED) evt(CMD_SCAN)
**trwifi_serialize_scaninfo: 171 [NETMGR][ERR]    malloc fail 0**
wifi_netmgr_utils_scan_ap: [RTK] WIFi Scan success
wifimgr_handle_request: 621 [WM][INFO]  handle request state(SCANNING) evt(EVT_SCAN_DONE)
[WT][T6] -->     _wt_scan_done:369
[WT][T6] send signal    _wt_scan_done:374
[WT][T17] wait signal   _wt_scan:578

We tried not to post "TRWIFI_POST_SCANEVENT(ameba_nm_dev_wlan0, LWNL_EVT_SCAN_DONE, (void *)g_scan_list);" in case of g_scan_list = NULL. 
But if we make the above changes, we will get the below result

TASH>>wm_test scan hello
TASH>>[WT][T14] wifi manager test!!     wm_test_main:907
[WT][T14] wait func signal      wm_test_main:914
[WT][T15] -->   _wt_scan:561
wifi_manager_scan_specific_ap: 197 [WM][INFO]   -->
wifimgr_handle_request: 621 [WM][INFO]  handle request state(STA_DISCONNECTED) evt(CMD_SCAN)
wifi_netmgr_utils_scan_ap: [RTK] WIFi Scan success
[WT][T15] wait signal   _wt_scan:578

**TASH>>wm_test scan AP002
TASH>>[WT][T16] wifi manager test!!     wm_test_main:907
[WT][ERR][T16] Program is already running       _wt_signal_init:331**

Please help to take a look at the above cases and let me know in case of any other suggestions.

**Comment by pillip8282** [comment]
> @pillip8282 , I updated the PR according to suggestions and would like to highlight two issues.

1) SSID with space: for eg: "JJ Residence"
Unfortunately TASH can't pass arguments contain space to program. So you have to pass the argument without tash. 
I added simple program to PR4984, and you can pass the argument by modifying TR_SSID in hello_main.c
You can test it with following command. 
TASH> hello 0

2)  For the case with scan_cnt=0:
Even scan_cnt is 0, if trwifi_scan_ap returned success, then the wi-fi driver has to generate callback event. 
typedef trwifi_result_e (*trwifi_scan_ap)(struct netdev *dev, trwifi_ap_config_s *config);
Because a wi-fi application decides to wait the signal depend on return of trwifi_scan_ap(). 
So if no AP is discovered then call TRWIFI_POST_SCANEVENT(dev, LWNL_EVT_SCAN_FAILED, NULL);

**Comment by keerthanalex** [comment]
> 2\. LWNL_EVT_SCAN_FAILED

@pillip8282 ,
1. SSID with space: for eg: "JJ Residence"
Since user can't pass the argument through "wm_test scan <SSID>" for SSID with space, we can exclude these cases from testing. Is that correct?

2. For the case with scan_cnt=0:
In case of scan_cnt = 0, we need to consider it as a failed case and call "TRWIFI_POST_SCANEVENT(dev, LWNL_EVT_SCAN_FAILED, NULL". Is that correct?

Do you have any review points on other part of implementation? so that I can update altogether.

**Comment by pillip8282** [comment]
> > 2. LWNL_EVT_SCAN_FAILED
> 
> @pillip8282 ,
> 
> 1. SSID with space: for eg: "JJ Residence"
>    Since user can't pass the argument through "wm_test scan " for SSID with space, we can exclude these cases from testing. Is that correct?
> 2. For the case with scan_cnt=0:
>    In case of scan_cnt = 0, if no AP is discovered "app_scan_result_handler" calls "TRWIFI_POST_SCANEVENT" with g_scan_list = NULL. The "trwifi_serialize_scaninfo()" API returns -1 with **"trwifi_serialize_scaninfo: 171 [NETMGR][ERR] malloc fail 0"** and then calls the "trwifi_post_event(dev, LWNL_EVT_SCAN_FAILED, NULL, 0);".

user can pass the argument. because user didn't pass it through TASH. So I made simple testcases PR4984 to test these cases.

> 
> Shouldn't we prevent "trwifi_serialize_scaninfo()" API from "malloc" when g_scan_list = NULL and directly calls "LWNL_EVT_SCAN_FAILED"?

you're right. I modified it to proceed trwifi_serialize_scaninfo if g_scan_list is not null.

And if you finish implementation. please let me know. I'll verify it.

**Comment by keerthanalex** [comment]
@pillip8282, I'm done with the implementation. please review and let me know in case of any modification.

**Review by pillip8282 — APPROVED** [review]
lgtm

**Pull request merged**